File: template_method/java/CuentaBancaria.java
package template_method.java;

public abstract class CuentaBancaria {

	protected String cbu;
	protected double saldo;

	public CuentaBancaria(
		String unCbu, double unSaldoInicial
	) {
		this.cbu = unCbu;
		this.saldo = unSaldoInicial;
	}

	//Método Plantilla - Template Method
	protected final void plantillaExtraer(double unMonto) {
		this.hacerValidar(unMonto);
		this.extraerGancho(unMonto);
		this.informarGancho(unMonto);
	}

	// Método abstracto puro redefinible por las subclases
	protected abstract boolean hacerValidar(double unMonto);

	// Método gancho con implementación por defecto
	protected void extraerGancho(double unMonto) {
		this.saldo -= unMonto;
	}

	// Método gancho con implementación por defecto
	protected void informarGancho(double unMonto) {
		System.out.println("Se extrajo con éxito de la cuenta " + this.cbu + " $" + unMonto);
	}
	
}

File: template_method/typescript/CuentaBancaria.ts
abstract class CuentaBancaria {

	protected cbu: string;
	protected saldo: number;

	public constructor(
		unCbu: string, unSaldo: number = 0
	) {
		this.cbu = unCbu;
		this.saldo = unSaldo;
	}

	//Método Plantilla - Template Method
	protected plantillaExtraer(unMonto: number): void {
		this.hacerValidar(unMonto);
		this.extraerGancho(unMonto);
		this.informarGancho(unMonto);
	}

	protected abstract hacerValidar(unMonto: number);

	// Método gancho con implementación por defecto
	protected extraerGancho(unMonto: number): void {
		this.saldo -= unMonto;
	}

	// Método gancho con implementación por defecto
	protected informarGancho(unMonto: number): void {
		console.log("Se extrajo con éxito de la cuenta " + this.cbu + " $" + unMonto);
	}

}